### PR TITLE
Context switch interception

### PR DESCRIPTION
**Pull request by pwnosaur**

### CODE REVIEW

**Comment by tklengyel** [comment]
I gave this PR a spin and after starting/stopping it a couple times the VM crashed. This suggest to me that the memory gets desynchronized and certain memory write events are not processed correctly. That would be a design issue with how the altp2m views are setup.

**Comment by pwnosaur** [comment]
> Never mind, looks like my VM just ran out of shadow memory. With that increased it works pretty well.

Great, I have marked it as ready.

**Comment by tklengyel** [comment]
@pwnosaur You still have ci checks failing

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

File: src/main.cpp [hunk 1]
```diff
@@ -331,6 +331,9 @@ int main(int argc, char** argv)
     int args_count = 0;
     bool terminate = false;
     int termination_timeout = 20;
+    bool context_based_interception = false;
+    char * context_processes [100] = {nullptr};
```
Comment (tklengyel) [inline src/main.cpp:335]: This should be a GSlist instead of an array

File: src/main.cpp [hunk 2]
```diff
@@ -455,6 +461,13 @@ int main(int argc, char** argv)
             case 'c':
                 inject_cwd = optarg;
                 break;
+            case 'C':
+                context_based_interception = true;
+                break;
+            case opt_context_interception_processes:
+                context_processes[context_processes_count] = optarg;
```
Comment (tklengyel) [inline src/main.cpp:468]: Here you would just prepend the `optarg` to the GSList

File: src/libdrakvuf/vmi.c [hunk 3]
```diff
@@ -1516,6 +1607,19 @@ bool init_vmi(drakvuf_t drakvuf, bool libvmi_conf, bool fast_singlestep)
     }
     PRINT_DEBUG("Altp2m view R created with ID %u\n", drakvuf->altp2m_idr);
 
+    /*
+     * IDRW View is used for context based interception, in order to protect
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:1611]: `idrx` you mean

File: src/libdrakvuf/vmi.c [hunk 4]
```diff
@@ -1365,6 +1423,39 @@ void drakvuf_loop(drakvuf_t drakvuf, bool (*is_interrupted)(drakvuf_t, void*), v
...
+        drakvuf->enable_cr3_based_interception = true;
+    }
+    else
+    {
+        status = vmi_slat_switch(drakvuf->vmi, drakvuf->altp2m_idx);
+        if (VMI_FAILURE == status)
+            PRINT_DEBUG("Disabling context based interception failed. \n");
+
+        if ( !drakvuf->cr3 )
+            control_cr3_trap(drakvuf, 0);
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:1451]: You shouldn't use `control_cr3_trap` directly. You should use and internal `drakvuf_trap` for that, so that this won't interfere with plugins that may also subscribe to cr3 traps.

File: src/libdrakvuf/vmi.c [hunk 5]
```diff
@@ -1162,18 +1214,24 @@ bool inject_trap_pa(drakvuf_t drakvuf,
...
     }
 
-    if ( !inject_trap_mem(drakvuf, &container->breakpoint.guard2, 1) )
+    if ( !inject_trap_mem(drakvuf, &container->breakpoint.guard2, 1, drakvuf->altp2m_idx) )
     {
         PRINT_DEBUG("Failed to create guard2 trap for the breakpoint!\n");
         goto err_exit;
     }
 
+    if ( !inject_trap_mem(drakvuf, &container->breakpoint.guard3, 0, drakvuf->altp2m_idrx) )
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:1229]: I would assume this is only needed if the context-based feature is used.

File: src/libdrakvuf/vmi.c [hunk 6]
```diff
@@ -684,7 +697,7 @@ event_response_t cr3_cb(vmi_instance_t vmi, vmi_event_t* event)
 
     flush_vmi(drakvuf);
 
-#ifdef DRAKVUF_DEBUG
+#ifndef DRAKVUF_DEBUG
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:700]: Unrelated change

File: src/libdrakvuf/vmi.c [hunk 7]
```diff
@@ -335,8 +341,16 @@ event_response_t post_mem_cb(vmi_instance_t vmi, vmi_event_t* event)
 done:
     free_proc_data_priv_2(&pass->proc_data, &pass->attached_proc_data);
     g_slice_free(struct memcb_pass, pass);
+
+    uint16_t view = drakvuf->altp2m_idx;
+
+    // Switch to RX
+    if(!drakvuf->vcpu_monitor[event->vcpu_id] && drakvuf->enable_cr3_based_interception)
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:348]: Reverse the order in the if

File: src/libdrakvuf/private.h [hunk 8]
```diff
@@ -325,6 +330,13 @@ struct memcb_pass
     GSList* traps;
 };
 
+typedef struct intercept_process
+{
+    char * name;
+    vmi_pid_t pid;
+    bool strict;
```
Comment (tklengyel) [inline src/libdrakvuf/private.h:337]: Please document what the `strict` field means

File: src/drakvuf.cpp [hunk 9]
```diff
@@ -364,6 +372,5 @@ void drakvuf_c::terminate(vmi_pid_t injection_pid,
 
     auto info = termination_info(terminated_processes, pid);
     drakvuf_loop(drakvuf, is_terminated, &info);
-
```
Comment (tklengyel) [inline src/drakvuf.cpp:367]: Unrelated change

File: src/drakvuf.cpp [hunk 10]
```diff
@@ -364,6 +372,5 @@ void drakvuf_c::terminate(vmi_pid_t injection_pid,
...
     cleanup_timer(this, timeout_thread);
-}
```
Comment (tklengyel) [inline src/drakvuf.cpp:369]: Unrelated change

File: src/libdrakvuf/vmi.c [hunk 11]
```diff
@@ -703,11 +716,36 @@ event_response_t cr3_cb(vmi_instance_t vmi, vmi_event_t* event)
     while (loop)
     {
         trap_info.trap = (drakvuf_trap_t*)loop->data;
+
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:719]: Unrelated change

File: src/libdrakvuf/vmi.c [hunk 12]
```diff
@@ -1121,6 +1160,13 @@ bool inject_trap_pa(drakvuf_t drakvuf,
             PRINT_DEBUG("%s: Failed to change gfn on view %u\n", __FUNCTION__, drakvuf->altp2m_idr);
             goto err_exit;
         }
+
+        if (VMI_FAILURE == vmi_slat_change_gfn(
+        drakvuf->vmi, drakvuf->altp2m_idrx, remapped_gfn->r, drakvuf->sink_page_gfn))
```
Comment (tklengyel) [inline src/libdrakvuf/vmi.c:1165]: I don't see the `sink_page_gfn` getting its `rwx` permissions removed in the `rx` view.

**Review by tklengyel — CHANGES_REQUESTED** [review]

File: src/libdrakvuf/vmi.c (hunk 5 above)
Comment (pwnosaur) [reply src/libdrakvuf/vmi.c:1229]: It could be used in case we allow the plugins to toggle the context-based feature, in that case the breakpoints has to be set in order to make the toggling process easier.

File: src/libdrakvuf/vmi.c (hunk 4 above)
Comment (pwnosaur) [reply src/libdrakvuf/vmi.c:1451]: I'm not sure if this would interfere with other subscribers as it does check if there is registered traps in `drakvuf->cr3` in case there is it wouldn't disable the cr3_event, and will only be disabled if the list is empty and there are no subscribers.

On the other hand , using `drakvuf_add_trap` would require a separate callback for the cr3_interception check, as well as in the process of handling a cr3 event, it would invoke the callback to its relative position in the GSList, which could be in the middle, in that case it may affect the response flags (not necessarily), tho it is your decision to make.

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by tklengyel** [comment]
Valgrind is spotting some errors now

File: src/libdrakvuf/vmi.c (hunk 4 above)
Comment (tklengyel) [reply src/libdrakvuf/vmi.c:1456]: I see, OK

**Review by tklengyel — APPROVED** [review]

**Comment by tklengyel** [comment]
I'll add a test to Jenkins to exercise this new mode and rerun the tests shortly

File: src/drakvuf.cpp [hunk 13]
```diff
@@ -286,6 +286,19 @@ void drakvuf_c::resume()
     drakvuf_resume(drakvuf);
 }
 
+void drakvuf_c::toggle_context_interception(GSList* processes)
+{
+    GSList* process = processes;
+
+    while (process != NULL)
+    {
+        drakvuf_intercept_process_add(this->drakvuf, (char*)process->data, 0, false);
```
Comment (tklengyel) [inline src/drakvuf.cpp:295]: It would be helpful if the user could specify either process name or pid from the command line. Right now only process name is possible. I don't think we need separate command-line options for this, if the input is just a number I think it would be safe to assume its a PID.

Comment (pwnosaur) [reply src/drakvuf.cpp:295]: Indeed it would be helpful, but rather than allowing the user to specify PID only instead of the name, It would be safer to allow the user to specify the PID alongside the name because it is more likely to have another process with the same PID in a short amount of time but less likely that it will have the same name and PID; therefore the argument for this option could be separated using a delimiter `process_name, PID`, and the user will have the option to specify either the name only or name & PID.

Comment (tklengyel) [reply src/drakvuf.cpp:295]: Sure, but I particularly want to be able to use just PID. Instead of comma-delimiter I would suggest something like `process-name:pid`. If no `:` is present treat it as process-name non-strict mode. If `:` is present but nothing before it, then use just pid, like `:1234`.

File: src/libdrakvuf/private.h [hunk 14]
```diff
@@ -334,7 +334,7 @@ typedef struct intercept_process
 {
     char* name;
     vmi_pid_t pid;
-    bool strict; /* True: Match both Name and PID, False: Match just the name */
+    unsigned char strict; /* 0: Match Name, 1: Match PID , 2: Match PID and Name */
```
Comment (tklengyel) [inline src/libdrakvuf/private.h:337]: This probably would be better to be defined as an enum

File: src/drakvuf.cpp [hunk 15]
```diff
@@ -296,18 +296,18 @@ void drakvuf_c::toggle_context_interception(GSList* processes)
...
 
         if (process_arg[0] == ':')
         {
             pid = atoi(&process_arg[1]);
-            strictness = 1;
+            strictness = MATCH_PID;
         }
         else
         {
             name = strtok(process_arg, ":");
```
Comment (tklengyel) [inline src/drakvuf.cpp:308]: `g_strsplit` please

Comment (pwnosaur) [reply src/drakvuf.cpp]: Done.

**Comment by tklengyel** [comment]
Thank you, this looks good! Much appreciated and very handy feature to have.
